File: Week 7/id_445/191.位-1-的个数.cpp
/*
 * @lc app=leetcode.cn id=191 lang=cpp
 *
 * [191] 位1的个数
 */
#include <stdint.h>
using namespace std;
// @lc code=start
class Solution {
public:
//solution 1
    // int hammingWeight(uint32_t n) {
    //     return (n > 0) ? 1 + hammingWeight(n & (n - 1)) : 0;
        
    // }
/// solution 2
//  int hammingWeight(uint32_t n) {
//      int count=0;
//      while (n>0)
//      {
//          n=n & (n - 1);
//          count++;
         
//      }
//      return count;
     
//         // return (n > 0) ? 1 + hammingWeight(n & (n - 1)) : 0;
        
//     }

    //solution 3
    int hammingWeight(uint32_t n) {
        int count=0;
        int mask=1;
        while (n>0)
        {
            if((n&mask)!=0) count++;
            n=n>>1;
        }
        return count;
        
    }
};
// @lc code=end

File: Week 7/id_520/191.位-1-的个数.java
/*
 * @lc app=leetcode.cn id=191 lang=java
 *
 * [191] 位1的个数
 */

// @lc code=start
public class Solution {
    // you need to treat n as an unsigned value
    public int hammingWeight(int n) {
        // 方法一，循环和位移动
        // int nums = 0, mask = 1;
        // for (int i = 0 ; i < 32; i++) {
        //     if ((n & mask) != 0) {
        //         nums++;
        //     } 
        //     mask <<= 1;
        // }
        // return nums;
        
        // 方法二， 位操作的小技巧
        int sum = 0;
        while (n != 0) {
            sum++;
            n &= (n - 1);
        }
        return sum;
    }
}
// @lc code=end

